### PR TITLE
Add support ORT whisper

### PR DESCRIPTION
**Pull request by mht-sharma**
# What does this PR do?

This PR enables the export of Whisper model to ONNX.

To enable this new modality, I had refactored the existing `ORTModelForConditionalGeneration` to add support for the multimodal models.

The PR has a dependency on transformers PR [19525](https://github.com/huggingface/transformers/pull/19525) which integrates the onnx config for the whisper model and adds support for the audio preprocessor.

# Usage

Using Transformers `AutoModelForSpeechSeq2Seq`
```python
>>> from datasets import load_dataset
>>> from transformers import AutoModelForSpeechSeq2Seq, AutoProcessor, pipeline

>>> processor = AutoProcessor.from_pretrained("openai/whisper-tiny.en")
>>> model = AutoModelForSpeechSeq2Seq.from_pretrained("openai/whisper-tiny.en")
>>> speech_recognition_pipeline = pipeline(
...    "automatic-speech-recognition",
...     model=model,
...     feature_extractor=processor.feature_extractor,
...     tokenizer=processor.tokenizer,
...  )

>>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
>>> inputs = processor(ds[0]["audio"]["array"], return_tensors="pt")
>>> result = speech_recognition_pipeline(ds[0]["audio"]["array"])
```

Using Optimum `ORTModelForSpeechSeq2Seq`
```python
>>> from datasets import load_dataset
>>> from transformers import AutoProcessor, pipeline
>>> from optimum.onnxruntime import ORTModelForSpeechSeq2Seq

>>> processor = AutoProcessor.from_pretrained("openai/whisper-tiny.en")
>>> model = ORTModelForSpeechSeq2Seq("openai/whisper-tiny.en", from_transformers=True)
>>> speech_recognition_pipeline = pipeline(
...    "automatic-speech-recognition",
...     model=model,
...     feature_extractor=processor.feature_extractor,
...     tokenizer=processor.tokenizer,
...  )

>>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
>>> inputs = processor(ds[0]["audio"]["array"], return_tensors="pt")
>>> result = speech_recognition_pipeline(ds[0]["audio"]["array"])
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (lewtun) [inline optimum/onnxruntime/modeling_seq2seq.py:741]: Same as above regarding `onnx_config` not being used

**Review by lewtun — COMMENTED** [review]
Thanks for working on this great feature @mht-sharma 🚀 !!!

The API looks very good to me and most of my comments are nits. Did you manage to resolve the issue with the generations not matching the `transformers` model?

File: optimum/onnx/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnx/modeling_seq2seq.py:68]: For models such as Whisper, there is no inputs named encoder_attention_mask which can result in error.

QQ.  The following `_DecoderWithLMhead` class could now be removed. Now the decoder exports work by passing the encoder_outputs as input. Hence, there is no need for a separate class to concatenate the decoder with lm head.

File: tests/onnxruntime/test_modeling.py
Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py:1216]: These 2 tests were not working yet so these are commented.

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py:733]: The functions were added to prevent any hardcoding based on model type in the `ORTModelForConditionalGeneration._from_transformers()` function for retrieving the onnx configs for encoder and decoder.

For the seq2seq models such as T5 the onnx_configs are generated using local Encoder and Decoder configs, hence this variable becomes unused.

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: This should not break anything. But added the default value for input name

**Comment by mht-sharma** [comment]
QQ - Do we run the CI on gpu machine? The following test case tests the pipeline on GPU [L1148](https://github.com/huggingface/optimum/blob/main/tests/onnxruntime/test_modeling.py#L1148). However, as per my understanding the pipeline is not running on GPU and it should fail or am I missing something.

**Comment by JingyaHuang** [comment]
> QQ - Do we run the CI on gpu machine? The following test case tests the pipeline on GPU [L1148](https://github.com/huggingface/optimum/blob/main/tests/onnxruntime/test_modeling.py#L1148). However, as per my understanding the pipeline is not running on GPU and it should fail or am I missing something.

Hey @mht-sharma, the [CI for GPU](https://github.com/huggingface/optimum/blob/main/.github/workflows/test_onnxruntime_gpu.yml) is scheduled to run nightly.

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [inline optimum/onnxruntime/modeling_seq2seq.py:851]: Currently,  the class `ORTModelForSpeechSeq2Seq` returns generic encoder / decoder ONNX configs. Same design is for the `ORTModelForSeq2SeqLM` class. 

`AutoModelForSpeechSeq2Seq` contains multiple speech model classes within it. There may be some models which are not supported using the generic ONNX configs (Need to confirm)? Similar to AutoClasses we may need a structure to switch to appropriate Onnx Config based on the model types. Probably something we can work after refactor.

File: optimum/onnx/configuration.py
Comment (lewtun) [inline optimum/onnx/configuration.py:362]: While reviewing Michael's ONNX PR, I realised we can take the opportunity to align our axis names with the same ones used in the `transformers` docs:

```suggestion
                ("input_features", {0: "batch_size", 1: "feature_size", 2: "encoder_sequence_length"}),
```

File: optimum/onnx/configuration.py
Comment (lewtun) [inline optimum/onnx/configuration.py:368]:
```suggestion
        return OrderedDict({"last_hidden_state": {0: "batch_size", 1: "encoder_sequence_length"}})
```

File: optimum/onnx/configuration.py
Comment (lewtun) [inline optimum/onnx/configuration.py:380]:
```suggestion
                ("input_ids", {0: "batch_size", 1: "past_decoder_sequence_length + sequence_length"}),
```

File: optimum/onnx/configuration.py
Comment (lewtun) [inline optimum/onnx/configuration.py:381]:
```suggestion
                ("encoder_hidden_states", {0: "batch_size", 1: "encoder_sequence_length"}),
```

File: optimum/onnx/configuration.py
Comment (lewtun) [inline optimum/onnx/configuration.py:389]: FYI this method will be superseded by @michaelbenayoun's ONNX PR in #403 

It's probably OK to keep things as they are for now and then refactor later IMO (unless there's objections from the team of course :))

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (lewtun) [reply optimum/onnxruntime/modeling_seq2seq.py:851]: Since the whole ONNX part is also changing, I think it's fine to allow time for a refactor to support audio if needed

File: tests/onnxruntime/test_modeling.py
Comment (lewtun) [inline tests/onnxruntime/test_modeling.py:1252]: remove?

**Review by lewtun — APPROVED** [review]
Thanks a lot for making this beast of a model work with `optimum` @mht-sharma !

Out of curiosity, is there any significant slow down when running the ORT model in CPU/GPU?

If not, I think this PR is in great shape - the main question is whether we should merge + refactor for the new ONNX exporter in #403 or wait until that PR is merged (at the expense of delaying this feature)?

WDYT @michaelbenayoun @echarlaix ?

**Comment by michaelbenayoun** [comment]
@lewtun I think it makes more sense to merge the `exporters` PR first, since it lays the foundation for the onnx export in optimum.
Once this is merged, we can refactor and add whisper as the first contribution to this.

File: optimum/onnx/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnx/modeling_seq2seq.py]: No longer relevant

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: Done

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: Not relevant

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py]: No longer required

File: optimum/onnx/configuration.py
Comment (mht-sharma) [reply optimum/onnx/configuration.py]: Updated

File: optimum/onnx/configuration.py
Comment (mht-sharma) [reply optimum/onnx/configuration.py]: Updated in exporters

File: optimum/onnx/configuration.py
Comment (mht-sharma) [reply optimum/onnx/configuration.py]: Updated in exporters

File: optimum/onnx/configuration.py
Comment (mht-sharma) [reply optimum/onnx/configuration.py]: Updated in exporters

File: optimum/onnx/configuration.py
Comment (mht-sharma) [reply optimum/onnx/configuration.py]: Refactored based on exporters

File: tests/onnxruntime/test_modeling.py
Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py]: Done

**Comment by mht-sharma** [comment]
The PR is now updated with the following changes.

1. `Whisper` configs have been added to `exporters` along with the `SpeechSeq2Seq` configs.
2. `modeling_seq2seq.py` in ORT is updated to support `Whisper` using the configs from `exporters`. The file has been hardcoded to support only `Whisper` through `exporters` as discussed with @michaelbenayoun .
3. IOBinding support is added. Currently there is a bug when running inference using `past-key-values` for BS>1 using iobinding. PR for the same will be added by @JingyaHuang  soon.

Gently pinging team members for review. @lewtun @michaelbenayoun

File: optimum/exporters/onnx/convert.py
Comment (mht-sharma) [inline optimum/exporters/onnx/convert.py:115]: The  will always return the same name of the input as was present in the original model. However, in some cases we may want to change the input name in the generated `ONNX` model.

For example: For the following config [SpeechSeq2SeqDecoderOnnxConfig](https://github.com/huggingface/optimum/pull/420/files#diff-cf7d52b6c0ee70bc45f40c2c40ac99b49dbf02f2fdb0a141e9483f764126e05eR709), it makes sense to have a different input names than the dummy inputs provided to the model for export.

Currently I have just added a temporary change to test the end-to-end pipeline. @michaelbenayoun could you suggest a permanent fix? One could be to revert back to using `OrderedDict` for `inputs`?

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [inline optimum/onnxruntime/modeling_seq2seq.py:789]: This way to get the output shape seems to be too specific for the `Whisper` model and may not fit well for other `SpeechSeq2seq` models? Is there a way to avoid giving output shapes? @JingyaHuang 

Or if this is the only way should we rename the class to `ORTModelForWhisperConditionalGeneration`? This may lead to have different classes for each model type in future. Probably something to discuss.

File: tests/onnxruntime/test_modeling.py
Comment (mht-sharma) [inline tests/onnxruntime/test_modeling.py:1469]: Preprocessing for audio file requires installation of `soundfile` package. Should this be added to `setup.py` file? Or should we just generate dummy data for testing?

File: optimum/exporters/onnx/base.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:123]: Is it the "official" name?

We could take:

- `automatic-speech-recognition` to match the pipelines
- `speech2text`

@lewtun wdty?

File: optimum/exporters/onnx/model_configs.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:658]: I think this can be added to `optimum/utils/input_generators.py` since it's kind of generic enough.

Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:658]:
```suggestion
class DummySeq2SeqDecoderTextInputGenerator(DummyTextInputGenerator):
```

File: optimum/exporters/onnx/model_configs.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:714]: Why is this one needed?
I am asking to know whether `Seq2SeqDecoderTextInputGenerator` could be merged to `DummyDecoderTextInputGenerator`? Or maybe it would not make sense, just raising the question.

File: optimum/exporters/onnx/model_configs.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:720]: Not sure this is needed anymore!

File: optimum/exporters/onnx/model_configs.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:733]: Why is this needed compared to before?

File: optimum/exporters/onnx/model_configs.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/model_configs.py:750]: Does it depend to `use_past` or not?

File: optimum/utils/input_generators.py
Comment (michaelbenayoun) [inline optimum/utils/input_generators.py:100]: I guess it's not needed if it does not provide anything new.

File: tests/onnxruntime/test_modeling.py
Comment (michaelbenayoun) [reply tests/onnxruntime/test_modeling.py:1469]: Both are good IMO. You can add `soundfile` to the `testing` extra.

File: tests/onnxruntime/test_modeling.py
Comment (michaelbenayoun) [inline tests/onnxruntime/test_modeling.py:1452]: How long are those tests?

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (JingyaHuang) [reply optimum/onnxruntime/modeling_seq2seq.py:789]: There is a way to avoid giving the output shape -> bind the output with OrtValue which will be the case for custom tasks #447 , but the flaw is that then you need to transfer ownership across frameworks which is something that we try to avoid. IMO, if you can infer the output shape, you shall bind it directly with a torch tensor.  cc. @philschmid

File: optimum/exporters/onnx/convert.py
Comment (mht-sharma) [reply optimum/exporters/onnx/convert.py:115]: As discussed would create a mapping between the input names to solve the issue

File: optimum/exporters/onnx/model_configs.py
Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py]: Added to input_generators in [af7ddd3](https://github.com/huggingface/optimum/pull/420/commits/af7ddd37c11033b6bcd4d654c3604f7737f8767d)

Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py]: Updated in [af7ddd3](https://github.com/huggingface/optimum/pull/420/commits/af7ddd37c11033b6bcd4d654c3604f7737f8767d)

File: optimum/exporters/onnx/model_configs.py
Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py:674]: The `encoder_outputs` would be generated only for exporting decoders present in the `seq2seq` models as the whole model (encoder+decoder) is passed as input to export such model. Hence I have created a separate class to differentiate between them.

File: optimum/exporters/onnx/model_configs.py
Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py]: Removed in [af7ddd3](https://github.com/huggingface/optimum/pull/420/commits/af7ddd37c11033b6bcd4d654c3604f7737f8767d)

File: optimum/exporters/onnx/model_configs.py
Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py:695]: The `ORTModel` requires the output to always return the hidden states hence this is added.

File: optimum/exporters/onnx/model_configs.py
Comment (mht-sharma) [reply optimum/exporters/onnx/model_configs.py:700]: Same reason as above. To always output `hidden_states` in output the following function is required. `Note: ` When `use_past` is `False` then the hidden states are present in the output but the input do not contain the `past_key_values`.

File: optimum/utils/input_generators.py
Comment (mht-sharma) [reply optimum/utils/input_generators.py]: yes removed

File: optimum/exporters/onnx/base.py
Comment (lewtun) [reply optimum/exporters/onnx/base.py:123]: I think the idea was to partially align with the underlying [autoclass](https://huggingface.co/docs/transformers/v4.24.0/en/model_doc/auto#transformers.AutoModelForSpeechSeq2Seq), but I agree `automatic-speech-recognition` would be more intuitive.

In general (not for this PR), I think we should take the opportunity to align more closely with the Hub tasks, e.g. `seq2seq-lm` could also be `text2text-generation` right?

File: optimum/exporters/onnx/convert.py
Comment (mht-sharma) [reply optimum/exporters/onnx/convert.py]: Updated in [af7ddd3](https://github.com/huggingface/optimum/pull/420/commits/af7ddd37c11033b6bcd4d654c3604f7737f8767d)

File: tests/onnxruntime/test_modeling.py
Comment (michaelbenayoun) [reply tests/onnxruntime/test_modeling.py:1452]: Great!

File: tests/onnxruntime/test_modeling.py
Comment (michaelbenayoun) [inline tests/onnxruntime/test_modeling.py:1455]: Can this be useful for the audio input generator?

Comment (mht-sharma) [reply tests/onnxruntime/test_modeling.py:1455]: This generates an audio signal however `DummyAudioGenerator` requires to generate preprocessed audio data which are Mel Spectrograms. We could add the entire process of preprocessing audio data in `DummyAudioGenerator` (similar to how feature_extractor is implemented), but it may be not be too much  useful for exporting the model.

File: optimum/exporters/onnx/base.py
Comment (michaelbenayoun) [reply optimum/exporters/onnx/base.py:123]: Alright then I guess we can keep `speech2seq-lm` for now since the other names are aligned to the AutoClass, and maybe change that (if needed) for all the tasks in another PR.

File: optimum/exporters/onnx/base.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:211]: I would make it clear that it is needed when the dummy input names and the exported input names do not match.

File: optimum/exporters/onnx/base.py
Comment (michaelbenayoun) [inline optimum/exporters/onnx/base.py:246]:
```suggestion
                name = self.torch_to_onnx_input_map.get(name, name)
```

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (michaelbenayoun) [reply optimum/onnxruntime/modeling_seq2seq.py:791]: Second this. We could have a general `model type -> ORTEncoder class` mapping, and `ORTModelForSpeechSeq2Seq` would use this:
```python
class ORTModelForSpeechSeq2Seq:
   ...
   ...

    def _initialize_encoder(
        self,
        session: onnxruntime.InferenceSession,
        config: transformers.PretrainedConfig,
        device: torch.device,
        use_io_binding: bool = True,
    ) -> ORTEncoderForSpeechSeq2Seq:
        return _MODEL_TYPE_TO_ORTENCODER.get(model_type, "default")(...)

File: optimum/exporters/onnx/base.py
Comment (mht-sharma) [reply optimum/exporters/onnx/base.py:211]: Updated

File: optimum/onnxruntime/modeling_seq2seq.py
Comment (mht-sharma) [reply optimum/onnxruntime/modeling_seq2seq.py:792]: Updated

**Comment by R4ZZ3** [comment]
Please change this line
model = ORTModelForSpeechSeq2Seq("openai/whisper-tiny.en", from_transformers=True)

-->

model = ORTModelForSpeechSeq2Seq.**from_pretrained**("openai/whisper-tiny.en", from_transformers=True)

**Comment by zara0m** [comment]
Hello, 

I have seen this [discussion ](https://huggingface.co/openai/whisper-large-v2/discussions/7#6398809b11095028d87b16a2) which shows how can we use pipeline for audios more than 30s and how to change task to transcribe; the chunk works for onnx too, but I couldn't change the task to transcribe in pipeline configuration for onnx. 
I appreciate it if you could help me solve this.

Thanks!

**Pull request merged**
